### DISCUSSION TITLE
Scope of the project

### DISCUSSION THREAD

**Discussion by iacore** [david-vanderson/dvui#12] [General]
The Web Platform has video, 2D canvas, flexbox/grid layout, typography, SVG...

As such, it's not easy for a UI library to be on par with the web. Even flutter struggles.

So I wonder, what's the planned scope of this project? For now, support for raster image is missing.

**Comment by david-vanderson** [comment]
Great question!  This project grew from a desire to use an immediate mode gui from zig for a podcast player application on a librem5 phone.  It also grew out of interface elements in a networked game I wrote.

I wanted something that could both be added to an existing application/game, but also used as the whole interface for an application.

The existing immediate-mode guis I looked into relied on decently high framerates for user interaction.  So part of this is to push how far you can go with the immediate-mode gui idea while processing input events in lower framerate situations.

While the web has lots of impressive technologies, it is very heavyweight to bundle into an application.  This is for people who don't need or want that tradeoff.

Planned scope includes:
- useable as the whole application UI (variable framerate)
- also can be bolted on to an existing application/game (don't own the event loop)
- works well with low and variable framerates
- useable in more resource constrained environments (maybe there will be a backend that doesn't produce triangles for rasterization)

The [podcast player](https://github.com/david-vanderson/podcast) is getting reasonably far along.  There's still a ton of stuff that needs to be added (including raster image support) but enough is already there to be useful.

Does this help answer your question?

**Reply by iacore** [reply]
yes

**Comment by david-vanderson** [comment]
What is interesting to you about this project?  Is there something you would like to do with it?

**Reply by iacore** [reply]
It being written in Zig is interesting. Maybe it's useful for making UI like PureData

I use egui (Rust) and imgui (C) for actual projects. I also heard of lvgl.

**Comment by iacore** [comment]
@david-vanderson  I have added "tabs" to my application using buttons.

not sure if a built-in tab widget should be in scope of the project. will need more time to find out.

https://github.com/david-vanderson/dvui/assets/74560659/8555de6c-98e2-4dc9-881c-9e64c7d422e8

**Reply by david-vanderson** [reply]
That looks pretty good!

@JosephABudd Did some horizontal (and vertical) tabs in https://github.com/JosephABudd/okp - I think they used essentially a horizontal scroll area with a menu inside with each tab being a menuItem (and some modifications to automatically highlight the selected tab).

I think either way is pretty good.  Do we need a specific tabview widget?  Maybe for convenience, but hopefully it is mostly combining existing things.

One thing people might want is tabs like you made them, but with a layout that moves to multiple rows if there's not enough space.  I should make a generic "CursorLayoutWidget" that lays out children in rows.  Is there a better name for the concept of laying out horizontally but dropping to the next row when you run out of space?

**Reply by iacore** [reply]
Flexbox, "Flexible box". You can just call them FlexLayout.

It is a superset of current box layout. Remember to support flow in any two directions (e.g. left to right + down to up).

**Reply by ghost** [reply]
My new tabs are shown in the readme at https://github.com/JosephABudd/kickzig. I offset the non-selected tabs for an added visual clue. The are taken from your menu code. Keeping up with your developing API has been a challenge but its worth it. I really like the changes that are being made.

**Reply by david-vanderson** [reply]
Very cool - the offset is a good idea!

Sorry about the API churn, you are definitely an early-adopter.  My impression is that the churn is getting less so at least the trend is good.  Thanks for sticking with it!

**Reply by ghost** [reply]
The API only gets better as it churns and writing the tab-bars was so much easier with the new API. If I had a better understanding of widgets I would have made a function for creating the bars.

Also clearing the dvui back-end in the main loop remedied the menu ghost issue. If I comment that line out the ghosts reappear so we know that fixed it.
